Add workflow for creating Kerchunk files for short-range forecasts

jpolchlo opened the pull request:
## Overview

Following on from #128, this PR offers a workflow to automate the process of creating Kerchunk files for short-range NWM forecasts.  This workflow creates a mirror of the current contents of s3://noaa-nwm-pds as a collection of individual Kerchunk JSON files, and then adds a combined Kerchunk file representing all the available data.  Each time this workflow runs, it takes a diff of the current state of the Amazon-provided source collection and only creates individual files for the newly-added sources.  Similarly, old JSON files that no longer have a matching netcdf source will be deleted.

At the moment, this works for the short-range forecasts, but in principle can be run for other forecast types with minimal alteration.  Ideally, this workflow will be run on a schedule, but I have yet to test that functionality.

- [x] Test workflow for function
- [x] Test workflow as a scheduled event
- [ ] Extend to other forecast types

Closes #129

### Checklist

- [x] ~~Ran `nbautoexport export .` in `/opt/src/notebooks` and committed the generated scripts. This is to make reviewing notebooks easier. (Note the export will happen automatically after saving notebooks from the Jupyter web app.)~~
- [x] ~~Documentation updated if needed~~
- [x] PR has a name that won't get you publicly shamed for vagueness

### Notes

This workflow is currently set to run on a `worker` node type.  This is because some early workflow tests caused a crash of one of the base nodes.  This has some implications for overall cluster function if this behavior tends to repeat.  Using a worker node adds a few minutes of lag to the process, as well as a small amount of extra cost.  If tests require multiple runs of this workflow, one can reserve a worker node for the duration of the testing window by applying the following manifest:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: hold-node
spec:
  nodeSelector:
    node-type: worker
  containers:
  - name: pause
    image: kubernetes/pause
```
Making sure to issue `kubectl delete pod hold-node` when testing is complete to release the reserved node.

## Testing Instructions

* Log in to argo.noaa.azavea.com
* Create a workflow using the YAML from this PR

rajadain commented:
http://argo.noaa.azavea.com/ doesn't load for me. Is it temporarily down?

rajadain commented:
Took about ~19 minutes total, and succeeded:

![image](https://github.com/azavea/noaa-hydro-data/assets/1430060/ce81a7df-7248-46bc-beea-95b4e953f32c)

jpolchlo commented:
Worth making a mention that @rajadain's timing may not be reflective of the overall runtime in practice.  When the cadence of execution is frequent, there should be only small updates to the individual JSON files, and the parallel clearing of stale files and individual kerchunk file creation should be very fast.  The combined Kerchunk step should take about 8 or 9 minutes at all times, if there is no existing worker node, the spin-up should take about 3 minutes.  There's probably also some extra time due to orchestration of the worker pods, but that ought to be de minimis (<1 minute).  Since the run illustrated above likely diffs from several days ago, that may be enough to pad out the runtime from the expected 13 or so minutes to the observed 19.

This also makes me realize that I should not run the combined Kerchunk process in the event that no changes were observed, as it would be an expensive no-op.

rajadain commented:
Something seems to be a bit off. The `feature_id`s in the combined kerchunk file don't seem to have any correspondence with the NHD COMIDs in the Lower Schuylkill HUC-10. They did have corresponding ones in this notebook: https://github.com/azavea/noaa-hydro-data/blob/master/src/nwm-predictions/7_evaluating_nwm_streamflow_predictions_kerchunk-combined.ipynb. Particularly, I'm talking about the heading **Subset the dataset to only the streams within the HUC**. In this case, we get no overlap:

![image](https://github.com/azavea/noaa-hydro-data/assets/1430060/4a8c49cb-3583-42da-b8e7-c3318fb50c6c)

Investigating ...

jpolchlo commented:
I've added an hourly updater for the combined and individual kerchunk files.  This runs as a CronWorkflow in Argo.  This is tested, and the cron job fires as expected, and it draws on the kerchunking workflow that has been updated to a template.  I can validate that the cron job sparked up, started a new worker node, ran to completion, deleted its pods, and closed down the worker node.  Total runtime was 00:18:26.  Speeding up the process would be very desirable.  The combined kerchunk file generation would be the target, since it accounts for more than half the runtime.  (For reference, it takes about 4 minutes for the node to come up (which inflates the timings of the `identify-diffs` step), under 4 minutes to create any new individual Kerchunk files (inflated due to an unusually long interval between the previous run leading to more files needing to be created), and about 10 minutes for the combined Kerchunk.)

@rajadain These correctness issues are concerning!  But I haven't yet figured out where they could be coming from, since the code that is in the workflow was sourced from the demo notebooks from #128.  It's for sure a possibility that there was a transcription error.  I'll have to look more closely.  Did you notice if the individual files work as expected?  It'd be worthwhile to know at which step the error was introduced.  I also wonder if some of this problem could be due to the existence of stale JSON files in the combined Kerchunk, which wasn't updated for a few days, and at last refresh had about a thousand stale references to forecast files.  It would be very weird if that were the reason, but it's a thing to consider.  I have to update the workflow to ignore the oldest day of forecasts to avoid this issue.

rajadain commented:
Tried this again today and I'm now seeing the right streams. Nice work!

rajadain commented:
Wanted to note that I'm seeing the failures again. Will address those as part of #130.

rajadain commented:
Can we please make the `combined.json` available in the `azavea-noaa-hydro-data-public` bucket, so that others may access it too?

rajadain commented:
Noting that using the `combined.json` being made in this workflow, I'm getting artifacts in the output:

![image](https://github.com/azavea/noaa-hydro-data/assets/1430060/ad3b3c6c-3f30-481e-84a8-64136c4cd69d)

Using a similar file made using my notebooks, I don't see them:

![image](https://github.com/azavea/noaa-hydro-data/assets/1430060/ef758563-9990-4c73-9e20-1608b38f7c19)

Investigating now ...

rajadain commented:
Two issues I'm seeing:

1. In the Jupyter Notebook version, there's more single files generated:
    ```bash
    aws --profile=noaa s3 ls s3://azavea-noaa-hydro-data/noaa-nwm-pds/nwm.20230617/short_range/ | wc -l
    432
    ```
    ```bash
    aws --profile=noaa s3 ls s3://azavea-noaa-hydro-data/kerchunk/nwm.20230617/short_range/ | wc -l
    252
    ```
    Notably, the Argo workflow is missing files after the `t13z` mark. I'm not sure why.
2. The files themselves are slightly different, although the difference doesn't feel material:
    ```diff
    --- nwm.20230617_jupyter/short_range/nwm.t00z.short_range.channel_rt.f001.conus.json	2023-06-22 10:49:12
    +++ nwm.20230617_argo/short_range/nwm.t00z.short_range.channel_rt.f001.conus.json	2023-06-22 10:49:04
    @@ -1,66 +1,63 @@
     {
         "version": 1,
    -    "templates": {
    -        "u": "s3://noaa-nwm-pds/nwm.20230617/short_range/nwm.t00z.short_range.channel_rt.f001.conus.nc"
    -    },
         "refs": {
    -        ".zgroup": "{\n    \"zarr_format\": 2\n}",
    -        ".zattrs": "{\n    \"Conventions\": \"CF-1.6\",\n    \"NWM_version_number\": \"v2.2\",\n    \"TITLE\": \"OUTPUT FROM NWM v2.2\",\n    \"cdm_datatype\": \"Station\",\n    \"code_version\": \"v5.2.0-beta2\",\n    \"dev\": \"dev_ prefix indicates development/internal meta data\",\n    \"dev_NOAH_TIMESTEP\": 3600,\n    \"dev_OVRTSWCRT\": 1,\n    \"dev_channelBucket_only\": 0,\n    \"dev_channel_only\": 0,\n    \"featureType\": \"timeSeries\",\n    \"model_configuration\": \"short_range\",\n    \"model_initialization_time\": \"2023-06-17_00:00:00\",\n    \"model_output_type\": \"channel_rt\",\n    \"model_output_valid_time\": \"2023-06-17_01:00:00\",\n    \"model_total_valid_times\": 18,\n    \"proj4\": \"+proj=lcc +units=m +a=6370000.0 +b=6370000.0 +lat_1=30.0 +lat_2=60.0 +lat_0=40.0 +lon_0=-97.0 +x_0=0 +y_0=0 +k_0=1.0 +nadgrids=@\",\n    \"station_dimension\": \"feature_id\",\n    \"stream_order_output\": 1\n}",
    -        "crs/.zarray": "{\n    \"chunks\": [],\n    \"compressor\": null,\n    \"dtype\": \"|S1\",\n    \"fill_value\": \"IA==\",\n    \"filters\": null,\n    \"order\": \"C\",\n    \"shape\": [],\n    \"zarr_format\": 2\n}",
    -        "crs/.zattrs": "{\n    \"_ARRAY_DIMENSIONS\": [],\n    \"_CoordinateAxes\": \"latitude longitude\",\n    \"esri_pe_string\": \"GEOGCS[\\\"GCS_WGS_1984\\\",DATUM[\\\"D_WGS_1984\\\",SPHEROID[\\\"WGS_1984\\\",6378137.0,298.257223563]],PRIMEM[\\\"Greenwich\\\",0.0],UNIT[\\\"Degree\\\",0.0174532925199433]];-400 -400 1000000000;-100000 10000;-100000 10000;8.98315284119521E-09;0.001;0.001;IsHighPrecision\",\n    \"grid_mapping_name\": \"latitude longitude\",\n    \"inverse_flattening\": 298.2572326660156,\n    \"long_name\": \"CRS definition\",\n    \"longitude_of_prime_meridian\": 0.0,\n    \"semi_major_axis\": 6378137.0,\n    \"semi_minor_axis\": 6356752.5,\n    \"spatial_ref\": \"GEOGCS[\\\"GCS_WGS_1984\\\",DATUM[\\\"D_WGS_1984\\\",SPHEROID[\\\"WGS_1984\\\",6378137.0,298.257223563]],PRIMEM[\\\"Greenwich\\\",0.0],UNIT[\\\"Degree\\\",0.0174532925199433]];-400 -400 1000000000;-100000 10000;-100000 10000;8.98315284119521E-09;0.001;0.001;IsHighPrecision\",\n    \"transform_name\": \"latitude longitude\"\n}",
    +        ".zgroup": "{\"zarr_format\":2}",
    +        ".zattrs": "{\"Conventions\":\"CF-1.6\",\"NWM_version_number\":\"v2.2\",\"TITLE\":\"OUTPUT FROM NWM v2.2\",\"cdm_datatype\":\"Station\",\"code_version\":\"v5.2.0-beta2\",\"dev\":\"dev_ prefix indicates development\\/internal meta data\",\"dev_NOAH_TIMESTEP\":3600,\"dev_OVRTSWCRT\":1,\"dev_channelBucket_only\":0,\"dev_channel_only\":0,\"featureType\":\"timeSeries\",\"model_configuration\":\"short_range\",\"model_initialization_time\":\"2023-06-17_00:00:00\",\"model_output_type\":\"channel_rt\",\"model_output_valid_time\":\"2023-06-17_01:00:00\",\"model_total_valid_times\":18,\"proj4\":\"+proj=lcc +units=m +a=6370000.0 +b=6370000.0 +lat_1=30.0 +lat_2=60.0 +lat_0=40.0 +lon_0=-97.0 +x_0=0 +y_0=0 +k_0=1.0 +nadgrids=@\",\"station_dimension\":\"feature_id\",\"stream_order_output\":1}",
    +        "crs/.zarray": "{\"chunks\":[],\"compressor\":null,\"dtype\":\"|S1\",\"fill_value\":\"IA==\",\"filters\":null,\"order\":\"C\",\"shape\":[],\"zarr_format\":2}",
    +        "crs/.zattrs": "{\"_ARRAY_DIMENSIONS\":[],\"_CoordinateAxes\":\"latitude longitude\",\"esri_pe_string\":\"GEOGCS[\\\"GCS_WGS_1984\\\",DATUM[\\\"D_WGS_1984\\\",SPHEROID[\\\"WGS_1984\\\",6378137.0,298.257223563]],PRIMEM[\\\"Greenwich\\\",0.0],UNIT[\\\"Degree\\\",0.0174532925199433]];-400 -400 1000000000;-100000 10000;-100000 10000;8.98315284119521E-09;0.001;0.001;IsHighPrecision\",\"grid_mapping_name\":\"latitude longitude\",\"inverse_flattening\":298.2572326660156,\"long_name\":\"CRS definition\",\"longitude_of_prime_meridian\":0.0,\"semi_major_axis\":6378137.0,\"semi_minor_axis\":6356752.5,\"spatial_ref\":\"GEOGCS[\\\"GCS_WGS_1984\\\",DATUM[\\\"D_WGS_1984\\\",SPHEROID[\\\"WGS_1984\\\",6378137.0,298.257223563]],PRIMEM[\\\"Greenwich\\\",0.0],UNIT[\\\"Degree\\\",0.0174532925199433]];-400 -400 1000000000;-100000 10000;-100000 10000;8.98315284119521E-09;0.001;0.001;IsHighPrecision\",\"transform_name\":\"latitude longitude\"}",
             "crs/0": "\u0000",
    -        "feature_id/.zarray": "{\n    \"chunks\": [\n        2776738\n    ],\n    \"compressor\": {\n        \"id\": \"zlib\",\n        \"level\": 2\n    },\n    \"dtype\": \"<i4\",\n    \"fill_value\": 0,\n    \"filters\": [\n        {\n            \"elementsize\": 4,\n            \"id\": \"shuffle\"\n        }\n    ],\n    \"order\": \"C\",\n    \"shape\": [\n        2776738\n    ],\n    \"zarr_format\": 2\n}",
    -        "feature_id/.zattrs": "{\n    \"_ARRAY_DIMENSIONS\": [\n        \"feature_id\"\n    ],\n    \"cf_role\": \"timeseries_id\",\n    \"comment\": \"NHDPlusv2 ComIDs within CONUS, arbitrary Reach IDs outside of CONUS\",\n    \"long_name\": \"Reach ID\"\n}",
    +        "feature_id/.zarray": "{\"chunks\":[2776738],\"compressor\":{\"id\":\"zlib\",\"level\":2},\"dtype\":\"<i4\",\"fill_value\":null,\"filters\":[{\"elementsize\":4,\"id\":\"shuffle\"}],\"order\":\"C\",\"shape\":[2776738],\"zarr_format\":2}",
    +        "feature_id/.zattrs": "{\"_ARRAY_DIMENSIONS\":[\"feature_id\"],\"cf_role\":\"timeseries_id\",\"comment\":\"NHDPlusv2 ComIDs within CONUS, arbitrary Reach IDs outside of CONUS\",\"long_name\":\"Reach ID\"}",
             "feature_id/0": [
    -            "{{u}}",
    +            "s3://noaa-nwm-pds/nwm.20230617/short_range/nwm.t00z.short_range.channel_rt.f001.conus.nc",
                 55895,
                 558203
             ],
    -        "nudge/.zarray": "{\n    \"chunks\": [\n        2776738\n    ],\n    \"compressor\": {\n        \"id\": \"zlib\",\n        \"level\": 2\n    },\n    \"dtype\": \"<i4\",\n    \"fill_value\": 0,\n    \"filters\": [\n        {\n            \"elementsize\": 4,\n            \"id\": \"shuffle\"\n        }\n    ],\n    \"order\": \"C\",\n    \"shape\": [\n        2776738\n    ],\n    \"zarr_format\": 2\n}",
    -        "nudge/.zattrs": "{\n    \"_ARRAY_DIMENSIONS\": [\n        \"feature_id\"\n    ],\n    \"_FillValue\": -999900,\n    \"add_offset\": 0.0,\n    \"coordinates\": \"latitude longitude\",\n    \"grid_mapping\": \"crs\",\n    \"long_name\": \"Amount of stream flow alteration\",\n    \"missing_value\": -999900,\n    \"scale_factor\": 0.009999999776482582,\n    \"units\": \"m3 s-1\",\n    \"valid_range\": [\n        -5000000,\n        5000000\n    ]\n}",
    +        "nudge/.zarray": "{\"chunks\":[2776738],\"compressor\":{\"id\":\"zlib\",\"level\":2},\"dtype\":\"<i4\",\"fill_value\":-999900,\"filters\":[{\"elementsize\":4,\"id\":\"shuffle\"}],\"order\":\"C\",\"shape\":[2776738],\"zarr_format\":2}",
    +        "nudge/.zattrs": "{\"_ARRAY_DIMENSIONS\":[\"feature_id\"],\"add_offset\":0.0,\"coordinates\":\"latitude longitude\",\"grid_mapping\":\"crs\",\"long_name\":\"Amount of stream flow alteration\",\"missing_value\":-999900,\"scale_factor\":0.009999999776482582,\"units\":\"m3 s-1\",\"valid_range\":[-5000000,5000000]}",
             "nudge/0": [
    -            "{{u}}",
    +            "s3://noaa-nwm-pds/nwm.20230617/short_range/nwm.t00z.short_range.channel_rt.f001.conus.nc",
                 2581782,
                 108587
             ],
    -        "qBtmVertRunoff/.zarray": "{\n    \"chunks\": [\n        2776738\n    ],\n    \"compressor\": {\n        \"id\": \"zlib\",\n        \"level\": 2\n    },\n    \"dtype\": \"<i4\",\n    \"fill_value\": 0,\n    \"filters\": [\n        {\n            \"elementsize\": 4,\n            \"id\": \"shuffle\"\n        }\n    ],\n    \"order\": \"C\",\n    \"shape\": [\n        2776738\n    ],\n    \"zarr_format\": 2\n}",
    -        "qBtmVertRunoff/.zattrs": "{\n    \"_ARRAY_DIMENSIONS\": [\n        \"feature_id\"\n    ],\n    \"_FillValue\": -9999000,\n    \"add_offset\": 0.0,\n    \"coordinates\": \"latitude longitude\",\n    \"grid_mapping\": \"crs\",\n    \"long_name\": \"Runoff from bottom of soil to bucket\",\n    \"missing_value\": -9999000,\n    \"scale_factor\": 0.0010000000474974513,\n    \"units\": \"m3\",\n    \"valid_range\": [\n        0,\n        20000000\n    ]\n}",
    +        "qBtmVertRunoff/.zarray": "{\"chunks\":[2776738],\"compressor\":{\"id\":\"zlib\",\"level\":2},\"dtype\":\"<i4\",\"fill_value\":-9999000,\"filters\":[{\"elementsize\":4,\"id\":\"shuffle\"}],\"order\":\"C\",\"shape\":[2776738],\"zarr_format\":2}",
    +        "qBtmVertRunoff/.zattrs": "{\"_ARRAY_DIMENSIONS\":[\"feature_id\"],\"add_offset\":0.0,\"coordinates\":\"latitude longitude\",\"grid_mapping\":\"crs\",\"long_name\":\"Runoff from bottom of soil to bucket\",\"missing_value\":-9999000,\"scale_factor\":0.0010000000474974513,\"units\":\"m3\",\"valid_range\":[0,20000000]}",
             "qBtmVertRunoff/0": [
    -            "{{u}}",
    +            "s3://noaa-nwm-pds/nwm.20230617/short_range/nwm.t00z.short_range.channel_rt.f001.conus.nc",
                 8666478,
                 4965066
             ],
    -        "qBucket/.zarray": "{\n    \"chunks\": [\n        2776738\n    ],\n    \"compressor\": {\n        \"id\": \"zlib\",\n        \"level\": 2\n    },\n    \"dtype\": \"<i4\",\n    \"fill_value\": 0,\n    \"filters\": [\n        {\n            \"elementsize\": 4,\n            \"id\": \"shuffle\"\n        }\n    ],\n    \"order\": \"C\",\n    \"shape\": [\n        2776738\n    ],\n    \"zarr_format\": 2\n}",
    -        "qBucket/.zattrs": "{\n    \"_ARRAY_DIMENSIONS\": [\n        \"feature_id\"\n    ],\n    \"_FillValue\": -999900000,\n    \"add_offset\": 0.0,\n    \"coordinates\": \"latitude longitude\",\n    \"grid_mapping\": \"crs\",\n    \"long_name\": \"Flux from gw bucket\",\n    \"missing_value\": -999900000,\n    \"scale_factor\": 9.999999747378752e-06,\n    \"units\": \"m3 s-1\",\n    \"valid_range\": [\n        0,\n        2000000000\n    ]\n}",
    +        "qBucket/.zarray": "{\"chunks\":[2776738],\"compressor\":{\"id\":\"zlib\",\"level\":2},\"dtype\":\"<i4\",\"fill_value\":-999900000,\"filters\":[{\"elementsize\":4,\"id\":\"shuffle\"}],\"order\":\"C\",\"shape\":[2776738],\"zarr_format\":2}",
    +        "qBucket/.zattrs": "{\"_ARRAY_DIMENSIONS\":[\"feature_id\"],\"add_offset\":0.0,\"coordinates\":\"latitude longitude\",\"grid_mapping\":\"crs\",\"long_name\":\"Flux from gw bucket\",\"missing_value\":-999900000,\"scale_factor\":9.999999747378752e-6,\"units\":\"m3 s-1\",\"valid_range\":[0,2000000000]}",
             "qBucket/0": [
    -            "{{u}}",
    +            "s3://noaa-nwm-pds/nwm.20230617/short_range/nwm.t00z.short_range.channel_rt.f001.conus.nc",
                 5216547,
                 3449931
             ],
    -        "qSfcLatRunoff/.zarray": "{\n    \"chunks\": [\n        2776738\n    ],\n    \"compressor\": {\n        \"id\": \"zlib\",\n        \"level\": 2\n    },\n    \"dtype\": \"<i4\",\n    \"fill_value\": 0,\n    \"filters\": [\n        {\n            \"elementsize\": 4,\n            \"id\": \"shuffle\"\n        }\n    ],\n    \"order\": \"C\",\n    \"shape\": [\n        2776738\n    ],\n    \"zarr_format\": 2\n}",
    -        "qSfcLatRunoff/.zattrs": "{\n    \"_ARRAY_DIMENSIONS\": [\n        \"feature_id\"\n    ],\n    \"_FillValue\": -999900000,\n    \"add_offset\": 0.0,\n    \"coordinates\": \"latitude longitude\",\n    \"grid_mapping\": \"crs\",\n    \"long_name\": \"Runoff from terrain routing\",\n    \"missing_value\": -999900000,\n    \"scale_factor\": 9.999999747378752e-06,\n    \"units\": \"m3 s-1\",\n    \"valid_range\": [\n        0,\n        2000000000\n    ]\n}",
    +        "qSfcLatRunoff/.zarray": "{\"chunks\":[2776738],\"compressor\":{\"id\":\"zlib\",\"level\":2},\"dtype\":\"<i4\",\"fill_value\":-999900000,\"filters\":[{\"elementsize\":4,\"id\":\"shuffle\"}],\"order\":\"C\",\"shape\":[2776738],\"zarr_format\":2}",
    +        "qSfcLatRunoff/.zattrs": "{\"_ARRAY_DIMENSIONS\":[\"feature_id\"],\"add_offset\":0.0,\"coordinates\":\"latitude longitude\",\"grid_mapping\":\"crs\",\"long_name\":\"Runoff from terrain routing\",\"missing_value\":-999900000,\"scale_factor\":9.999999747378752e-6,\"units\":\"m3 s-1\",\"valid_range\":[0,2000000000]}",
             "qSfcLatRunoff/0": [
    -            "{{u}}",
    +            "s3://noaa-nwm-pds/nwm.20230617/short_range/nwm.t00z.short_range.channel_rt.f001.conus.nc",
                 4807143,
                 409404
             ],
    -        "reference_time/.zarray": "{\n    \"chunks\": [\n        1\n    ],\n    \"compressor\": {\n        \"id\": \"zlib\",\n        \"level\": 2\n    },\n    \"dtype\": \"<i4\",\n    \"fill_value\": 0,\n    \"filters\": [\n        {\n            \"elementsize\": 4,\n            \"id\": \"shuffle\"\n        }\n    ],\n    \"order\": \"C\",\n    \"shape\": [\n        1\n    ],\n    \"zarr_format\": 2\n}",
    -        "reference_time/.zattrs": "{\n    \"_ARRAY_DIMENSIONS\": [\n        \"reference_time\"\n    ],\n    \"long_name\": \"model initialization time\",\n    \"standard_name\": \"forecast_reference_time\",\n    \"units\": \"minutes since 1970-01-01 00:00:00 UTC\"\n}",
    +        "reference_time/.zarray": "{\"chunks\":[1],\"compressor\":{\"id\":\"zlib\",\"level\":2},\"dtype\":\"<i4\",\"fill_value\":null,\"filters\":[{\"elementsize\":4,\"id\":\"shuffle\"}],\"order\":\"C\",\"shape\":[1],\"zarr_format\":2}",
    +        "reference_time/.zattrs": "{\"_ARRAY_DIMENSIONS\":[\"reference_time\"],\"long_name\":\"model initialization time\",\"standard_name\":\"forecast_reference_time\",\"units\":\"minutes since 1970-01-01 00:00:00 UTC\"}",
             "reference_time/0": "base64:eF5TYFnLCAAB6wDT",
    -        "streamflow/.zarray": "{\n    \"chunks\": [\n        2776738\n    ],\n    \"compressor\": {\n        \"id\": \"zlib\",\n        \"level\": 2\n    },\n    \"dtype\": \"<i4\",\n    \"fill_value\": 0,\n    \"filters\": [\n        {\n            \"elementsize\": 4,\n            \"id\": \"shuffle\"\n        }\n    ],\n    \"order\": \"C\",\n    \"shape\": [\n        2776738\n    ],\n    \"zarr_format\": 2\n}",
    -        "streamflow/.zattrs": "{\n    \"_ARRAY_DIMENSIONS\": [\n        \"feature_id\"\n    ],\n    \"_FillValue\": -999900,\n    \"add_offset\": 0.0,\n    \"coordinates\": \"latitude longitude\",\n    \"grid_mapping\": \"crs\",\n    \"long_name\": \"River Flow\",\n    \"missing_value\": -999900,\n    \"scale_factor\": 0.009999999776482582,\n    \"units\": \"m3 s-1\",\n    \"valid_range\": [\n        0,\n        5000000\n    ]\n}",
    +        "streamflow/.zarray": "{\"chunks\":[2776738],\"compressor\":{\"id\":\"zlib\",\"level\":2},\"dtype\":\"<i4\",\"fill_value\":-999900,\"filters\":[{\"elementsize\":4,\"id\":\"shuffle\"}],\"order\":\"C\",\"shape\":[2776738],\"zarr_format\":2}",
    +        "streamflow/.zattrs": "{\"_ARRAY_DIMENSIONS\":[\"feature_id\"],\"add_offset\":0.0,\"coordinates\":\"latitude longitude\",\"grid_mapping\":\"crs\",\"long_name\":\"River Flow\",\"missing_value\":-999900,\"scale_factor\":0.009999999776482582,\"units\":\"m3 s-1\",\"valid_range\":[0,5000000]}",
             "streamflow/0": [
    -            "{{u}}",
    +            "s3://noaa-nwm-pds/nwm.20230617/short_range/nwm.t00z.short_range.channel_rt.f001.conus.nc",
                 614098,
                 1967684
             ],
    -        "time/.zarray": "{\n    \"chunks\": [\n        1024\n    ],\n    \"compressor\": {\n        \"id\": \"zlib\",\n        \"level\": 2\n    },\n    \"dtype\": \"<i4\",\n    \"fill_value\": 0,\n    \"filters\": [\n        {\n            \"elementsize\": 4,\n            \"id\": \"shuffle\"\n        }\n    ],\n    \"order\": \"C\",\n    \"shape\": [\n        1\n    ],\n    \"zarr_format\": 2\n}",
    -        "time/.zattrs": "{\n    \"_ARRAY_DIMENSIONS\": [\n        \"time\"\n    ],\n    \"long_name\": \"valid output time\",\n    \"standard_name\": \"time\",\n    \"units\": \"minutes since 1970-01-01 00:00:00 UTC\",\n    \"valid_max\": 28117080,\n    \"valid_min\": 28116060\n}",
    +        "time/.zarray": "{\"chunks\":[1024],\"compressor\":{\"id\":\"zlib\",\"level\":2},\"dtype\":\"<i4\",\"fill_value\":null,\"filters\":[{\"elementsize\":4,\"id\":\"shuffle\"}],\"order\":\"C\",\"shape\":[1],\"zarr_format\":2}",
    +        "time/.zattrs": "{\"_ARRAY_DIMENSIONS\":[\"time\"],\"long_name\":\"valid output time\",\"standard_name\":\"time\",\"units\":\"minutes since 1970-01-01 00:00:00 UTC\",\"valid_max\":28117080,\"valid_min\":28116060}",
             "time/0": "base64:eF7t0AENAAAIwzBIkIlFPCLknYN1SwQIpApM6rhvAgTqGBAgECvQsefGCRB4bKUBDw==",
    -        "velocity/.zarray": "{\n    \"chunks\": [\n        2776738\n    ],\n    \"compressor\": {\n        \"id\": \"zlib\",\n        \"level\": 2\n    },\n    \"dtype\": \"<i4\",\n    \"fill_value\": 0,\n    \"filters\": [\n        {\n            \"elementsize\": 4,\n            \"id\": \"shuffle\"\n        }\n    ],\n    \"order\": \"C\",\n    \"shape\": [\n        2776738\n    ],\n    \"zarr_format\": 2\n}",
    -        "velocity/.zattrs": "{\n    \"_ARRAY_DIMENSIONS\": [\n        \"feature_id\"\n    ],\n    \"_FillValue\": -999900,\n    \"add_offset\": 0.0,\n    \"coordinates\": \"latitude longitude\",\n    \"grid_mapping\": \"crs\",\n    \"long_name\": \"River Velocity\",\n    \"missing_value\": -999900,\n    \"scale_factor\": 0.009999999776482582,\n    \"units\": \"m s-1\",\n    \"valid_range\": [\n        0,\n        5000000\n    ]\n}",
    +        "velocity/.zarray": "{\"chunks\":[2776738],\"compressor\":{\"id\":\"zlib\",\"level\":2},\"dtype\":\"<i4\",\"fill_value\":-999900,\"filters\":[{\"elementsize\":4,\"id\":\"shuffle\"}],\"order\":\"C\",\"shape\":[2776738],\"zarr_format\":2}",
    +        "velocity/.zattrs": "{\"_ARRAY_DIMENSIONS\":[\"feature_id\"],\"add_offset\":0.0,\"coordinates\":\"latitude longitude\",\"grid_mapping\":\"crs\",\"long_name\":\"River Velocity\",\"missing_value\":-999900,\"scale_factor\":0.009999999776482582,\"units\":\"m s-1\",\"valid_range\":[0,5000000]}",
             "velocity/0": [
    -            "{{u}}",
    +            "s3://noaa-nwm-pds/nwm.20230617/short_range/nwm.t00z.short_range.channel_rt.f001.conus.nc",
                 2690369,
                 2116774
             ]
    \ No newline at end of file
    ```
    Despite the difference being only in formatting, I think we should check what version of `kerchunk` we're using in Argo. I'm using `0.0.6` in Jupyter. `0.1.2` is the latest, I should try using that too.

rajadain commented:
After upgrading my Jupyter notebook to Kerchunk `0.1.2`, the JSON differences go away, and the correct behavior remains. It is thus coming from the missing files. Looking into how to fix that.

rajadain commented:
Alright I've figured it out. Each Prediction Short Range data bucket has ~1728 entries:

```bash
aws s3 ls s3://noaa-nwm-pds/nwm.20230524/short_range/ | wc -l
1728
```

However, the `list_objects_v2` we're using here has [a hard limit of only fetching 1000 keys](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/list_objects_v2.html), which I've tried overriding to no avail.

This doesn't happen in my notebooks because I use `s3fs`:

```python
fs = s3fs.S3FileSystem()
folders = ['nwm.20230617', 'nwm.20230618', 'nwm.20230619']

netcdf_urls = [f's3://noaa-nwm-pds/{folder}/short_range/nwm.t*z.short_range.channel_rt.*.conus.nc' for folder in folders]
netcdfs = [netcdf for netcdf_url in netcdf_urls for netcdf in fs.glob(netcdf_url)]
```

Is it possible to use `s3fs` here?

rajadain commented:
The updates are working! https://github.com/azavea/noaa-hydro-data/blob/tt/add-argo-consuming-notebook/src/nwm-predictions/8_evaluating_nwm_streamflow_predictions_kerchunk-combined_via-argo.ipynb

![image](https://github.com/azavea/noaa-hydro-data/assets/1430060/5823c5cb-8a31-4b7b-853e-b47a0073dad2)

jpolchlo commented:
There are a number of other possible changes/improvements to this PR.  But for the sake of wrapping up, I'll merge this as is.  We can file future PRs to make modifications.
